### DISCUSSION TITLE
What's the idiomatic way to store something for use for the lifetime of a request?

### DISCUSSION THREAD

**Discussion by rsslldnphy** [vercel/next.js#58842] [Help] [answered]
### Summary

Hi! I'm trying to get my head around the mental model of NextJS and its app router. Something I've done in previous server-side apps is store things I only want to create/calculate once per request on the request object, so it is available to any functions during the lifetime of that request that take the request object.

As a concrete example, I'm writing a multitenant app that pulls the tenant slug from a url path param. Previously, I had middleware that would pull the user id from the session and the slug from the url, then look up in the database whether the user had access to that organisation. I'd then store the organisation details on the request object to make them available to whatever request handlers get invoked.

I'm trying to figure out how to do something similar in NextJS and struggling to know where to start. What's the idiomatic way to do this? The best approach I've come up with so far is creating some kind of `getRequestContext` function that creates all this stuff, calling it at the top of every page component, and passing it in as the first argument to any other functions (like data fetchers etc) that they call. Is there a better way?

**Comment by philwolstenholme** [comment, marked as answer]
Sounds like a job for `cache()` potentially. This is a recent addition to React but is available for us to use in Next: https://react.dev/reference/react/cache

There are a few examples of Vercel repos using it, like this one: https://github.com/vercel/nextjs-subscription-payments/blob/c7867b2d9e08d033056293d12aeb9825b8331806/app/supabase-server.ts#L6-L11

**Reply by rsslldnphy** [reply]
Thanks! That example is interesting - the cached function appears to use the value of `cookies` that’s in lexical scope, but doesn’t take it as an argument. Based on my understanding of the docs, does that not mean that the function will keep returning the same value even if `cookies` has changed? I still don’t have a good mental model of NextJS, would these top level defs be recalculated for every request? Are the caches scoped to just one render cycle?

**Reply by rsslldnphy** [reply]
Just seen that it does, in fact, say that the cache is invalidated on each request. This could be exactly what I need, thanks!
